op-challenger/fault/factory_test.go

```diff
@@ -0,0 +1,189 @@
+package fault
+
+import (
+	"context"
+	"errors"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	gameCountErr = errors.New("game count error")
+	gameIndexErr = errors.New("game index error")
+)
+
+// TestGameLoader_FetchAllGames tests that the game loader correctly fetches all games.
+func TestGameLoader_FetchAllGames(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name        string
+		caller      *mockMinimalDisputeGameFactoryCaller
+		expectedErr error
+		expectedLen int
+	}{
+		{
+			name:        "success",
+			caller:      newMockMinimalDisputeGameFactoryCaller(10, false, false),
+			expectedErr: nil,
+			expectedLen: 10,
+		},
+		{
+			name:        "game count error",
+			caller:      newMockMinimalDisputeGameFactoryCaller(10, true, false),
+			expectedErr: gameCountErr,
+			expectedLen: 0,
+		},
+		{
+			name:        "game index error",
+			caller:      newMockMinimalDisputeGameFactoryCaller(10, false, true),
+			expectedErr: gameIndexErr,
+			expectedLen: 0,
+		},
+		{
+			name:        "no games",
+			caller:      newMockMinimalDisputeGameFactoryCaller(0, false, false),
+			expectedErr: nil,
+			expectedLen: 0,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			loader := NewGameLoader(test.caller)
+
+			games, err := loader.FetchAllGames(context.Background())
+			require.Equal(t, test.expectedErr, err)
+			require.Len(t, games, test.expectedLen)
+			expectedGames := test.caller.games
+			if test.expectedErr != nil {
+				expectedGames = make([]struct {
+					proxy     common.Address
+					timestamp *big.Int
+				}, 0)
+			}
+			require.ElementsMatch(t, expectedGames, translateGames(games))
+		})
+	}
+}
+
+func generateMockGames(count uint64) []struct {
+	proxy     common.Address
+	timestamp *big.Int
+} {
+	games := make([]struct {
+		proxy     common.Address
+		timestamp *big.Int
```

Comment:
You could simplify this test a lot by using `FaultDisputeGame` type instead of redeclaring the `struct` everywhere except the return type for the mock's `GameAtIndex` return value. Avoids a heap of duplication and copying etc.